### PR TITLE
V5.0.x: ofi: NIC selection update (revised)

### PR DESCRIPTION
**Pull request by amirshehataornl**
Port of the distances patches to OMPI 5.0.x release

### CODE REVIEW

**Comment by naughtont3** [comment]
Revised PR for #11645

**Comment by amirshehataornl** [comment]
This PR is meant to carry all the patches for the distances. So it will wait until 11728 lands on master, then I'll include it in the port. This way everything works.

**Comment by amirshehataornl** [comment]
> > This PR is meant to carry all the patches for the distances. So it will wait until 11728 lands on master, then I'll include it in the port. This way everything works.
> 
> Gotcha. The change was just merged to main - it is needed to maintain backward compatibility for EFA, so I really appreciate if you could include those 2 commits as well.
> 
> Or I can do it separately. let me know1

done

File: opal/mca/common/ofi/common_ofi.c [hunk 1]
```diff
@@ -445,63 +446,227 @@ static int check_provider_attr(struct fi_info *provider_info, struct fi_info *pr
...
-    obj = hwloc_get_non_io_ancestor_obj(topology, obj);
-    if (NULL != obj) {
-        result = hwloc_bitmap_intersects(proc_cpuset, obj->cpuset);
+    *num_distances = idx;
+
+out:
+    *valin = val;
+    return nearest;
+}
+
```
Comment (wenduwan) [inline opal/mca/common/ofi/common_ofi.c:588]: Curious why `#if HWLOC_API_VERSION < 0x00020000` was removed in this change?

e.g. https://github.com/open-mpi/ompi/blob/main/opal/mca/common/ofi/common_ofi.c#LL608C1-L608C35

**Comment by wenduwan** [comment]
Encountered another bug. Need to add this PR as well https://github.com/open-mpi/ompi/pull/11744

File: opal/mca/common/ofi/common_ofi.c (hunk 1 above)
Comment (amirshehataornl) [reply opal/mca/common/ofi/common_ofi.c:588]: > Curious why `#if HWLOC_API_VERSION < 0x00020000` was removed in this change?
> 
> e.g. https://github.com/open-mpi/ompi/blob/main/opal/mca/common/ofi/common_ofi.c#LL608C1-L608C35

Looks like I missed porting 
"Don't use hwloc v2 structures unless we have that version or above"
Trying to do that now

**Comment by amirshehataornl** [comment]
PR https://github.com/open-mpi/ompi/pull/11739/ Updated

**Comment by gpaulsen** [comment]
bot:ibm:retest

**Comment by gpaulsen** [comment]
`GitHub Action CI / Milestone Pull Request (pull_request_target)` had some issues.  Trying to retest that:

bot:aws:retest

**Comment by gpaulsen** [comment]
@wenduwan and @amirshehataornl please review, and we'll get it in.

**Comment by wenduwan** [comment]
@gpaulsen Yes I plan to test out the change today/tomo and report back.

**Comment by wenduwan** [comment]
@amirshehataornl  Got a compile error with internal (old) hwloc and libevent
```
make[2]: Entering directory `/home/ec2-user/ompi/opal/mca/timer'
  CC       base/timer_base_open.lo
base/timer_base_open.c: In function 'opal_timer_base_close':
base/timer_base_open.c:60:9: warning: unused variable 'ret' [-Wunused-variable]
     int ret;
         ^~~
  CCLD     libmca_timer.la
make[3]: Entering directory `/home/ec2-user/ompi/opal/mca/timer'
make[3]: Nothing to be done for `install-exec-am'.
make[3]: Leaving directory `/home/ec2-user/ompi/opal/mca/timer'
make[2]: Leaving directory `/home/ec2-user/ompi/opal/mca/timer'
Making install in mca/common/ofi
make[2]: Entering directory `/home/ec2-user/ompi/opal/mca/common/ofi'
  CC       common_ofi.lo
  LN_S     libopen-palmca_common_ofi.la
common_ofi.c: In function 'get_package_rank':
common_ofi.c:782:20: error: 'ranks_on_package' undeclared (first use in this function)
             return ranks_on_package;
                    ^~~~~~~~~~~~~~~~
common_ofi.c:782:20: note: each undeclared identifier is reported only once for each function it appears in
make[2]: *** [common_ofi.lo] Error 1
make[2]: Leaving directory `/home/ec2-user/ompi/opal/mca/common/ofi'
```

Need to pick up https://github.com/open-mpi/ompi/pull/11693 as well

**Comment by wenduwan** [comment]
I cherry picked 9046276bdfd9f42ac9cdfb6c4a19e66314cb29de and git rebased - ran through a couple 2 nodes tests and the change looks good.

@amirshehataornl Thank you!

**Comment by awlauria** [comment]
Is this ready to merge?

**Comment by amirshehataornl** [comment]
It looks good from my side.

**Comment by wenduwan** [comment]
@gpaulsen I think @amirshehataornl missed https://github.com/open-mpi/ompi/commit/9046276bdfd9f42ac9cdfb6c4a19e66314cb29de

I will open a PR to cherry-pick that change

**Comment by wenduwan** [comment]
The `current_package_rank` variable is undefined https://github.com/open-mpi/ompi/blob/v5.0.x/opal/mca/common/ofi/common_ofi.c#L786

**Comment by wenduwan** [comment]
@amirshehataornl Sorry I meant `ranks_on_package` [here](https://github.com/open-mpi/ompi/blob/v5.0.x/opal/mca/common/ofi/common_ofi.c#L782)

**Comment by amirshehataornl** [comment]
hmmm how did that build in the first place? Are you creating a PR to fix that, or should I?

**Comment by wenduwan** [comment]
Unfortunately Open MPI CI does not build OFI so this was not caught.

I have opened a backport PR https://github.com/open-mpi/ompi/pull/11775

**Pull request merged**